s3/src/s3_lib.py

```diff
@@ -256,18 +256,33 @@ def _on_relation_changed_event(self, event: RelationChangedEvent) -> None:
         if any(newval for newval in diff.added if self.relation_data._is_secret_field(newval)):
             self.relation_data._register_secrets_to_relation(event.relation, diff.added)
 
-        provider_lib_version = float(
-            self.relation_data.fetch_relation_field(event.relation.id, S3_LIB_VERSION_FIELD) or "0"
-        )
-        if provider_lib_version < 1 and not self.relation_data.fetch_my_relation_field(
+        is_provider_v0 = False
+        provider_data = self.relation_data.fetch_relation_data([event.relation.id])[
+            event.relation.id
+        ]
+        if len(provider_data) > 0 and S3_LIB_VERSION_FIELD not in provider_data:
+            # This means that provider has written something on its part of relation data,
+            # but that something is not its version -- this means provider will never write its version
+            # because that's something the provider is meant to write first (on relation-created)!!!
+            is_provider_v0 = True
+        elif (
+            S3_LIB_VERSION_FIELD in provider_data
+            and float(provider_data[S3_LIB_VERSION_FIELD]) < 1
+        ):
```

Comment:
```suggestion
        if len(provider_data) > 0 and S3_LIB_VERSION_FIELD not in provider_data:
            # This means that provider has written something on its part of relation data,
            # but that something is not its version -- this means provider will never write its version
            # because that's something the provider is meant to write first (on relation-created)!!!
        or (
            S3_LIB_VERSION_FIELD in provider_data
            and float(provider_data[S3_LIB_VERSION_FIELD]) < 1
        ):
```

**niptick** wouldn't this work?